core/state/database.go

```diff
@@ -186,8 +186,9 @@ func (db *CachingDB) Reader(stateRoot common.Hash) (Reader, error) {
 			readers = append(readers, newFlatReader(snap))
 		}
 	} else {
-		// If standalone state snapshot is not available, try to construct
-		// the state reader with database.
+		// If standalone state snapshot is not available (path scheme
+		// or the state snapshot is explicitly disabled in hash mode),
+		// try to construct the state reader with database.
 		reader, err := db.triedb.StateReader(stateRoot)
 		if err == nil {
```

Comment:
Afaict this only works because `hashdb.StateReader` returns an error, otherwise we would create two readers for a hashdb. Right now the logic (afaict) is:

```
if snapshot { 
     // create snapshot reader
} else {
    if hashdb {
        // do nothing
    } else {
        // create pathdb snapshot reader
    }
}
```
Which seems a bit convoluted, why not make it like this:
```
if triedb.Scheme() == rawdb.HashScheme && db.snap != nil {
    // construct legacy snap reader
}
if triedb.Scheme() == rawdb.PathScheme {
    // construct triedb.StateReader
}
```